src/dbnode/storage/series/buffer.go

```diff
@@ -241,21 +243,17 @@ func (b *dbBuffer) Tick(blockStates map[xtime.UnixNano]BlockState) bufferTickRes
 		buckets.removeBucketsUpToVersion(blockState.Version)
 
 		if buckets.streamsLen() == 0 {
+			t := tNano.ToTime()
 			// All underlying buckets have been flushed successfully, so we can
 			// just remove the buckets from the bucketsMap.
-
-			// TODO(juchan): in order to support cold writes, the buffer needs
-			// to tell the series that these buckets were evicted from the
-			// buffer so that the cached block in the series can either:
-			//   1) be evicted, or
-			//   2) be merged with the new data.
-			// This needs to happen because the buffer just flushed new data
-			// to disk, which the cached block does not have, and is
-			// therefore invalid. This was fine while the missing data was still
-			// in memory, but once we evict it from the buffer we need to make
-			// sure that we bust the cache as well.
-			b.removeBucketVersionsAt(tNano.ToTime())
-			evictedBuckets++
+			b.removeBucketVersionsAt(t)
+			// Pass which bucket got evicted from the buffer to the series.
+			// Data gets read in order of precedence: buffer -> cache -> disk.
+			// After a bucket gets removed from the buffer, data from the cache
+			// will be served. However, since data just got persisted to disk,
+			// the cached block is now stale, therefore we need to evict that
+			// block from cache so that the new data can be retrieved from disk.
```

Comment:
Can you remind me why we settled on evicting instead of merging? Its just easier right? If so might be nice to leave a comment here thats like: "Instead of evicting, we could perform a merge instead, but that would complicate the code for XYZ reasons"

Reply:
Right. I think it's unclear which one is "more correct" in this circumstance, so it makes sense to do the cleaner, easier one. I'll put in a comment.